### PR TITLE
Add Image Embedding (Features) Vector to Output Results

### PR DESCRIPTION
**Pull request by davidamacey**
### Summary
🧠🔮🌈

This pull request adds support for using and storing the backbone embedding for different tasks in the `ultralytics` framework. The embedding is computed and normalized by the `BaseModel` class and passed to the `BasePredictor` and `Results` classes for each task. This enables new functionalities and improvements based on image embeddings, such as similarity search, clustering, or real-time detection and tracking.

> _Come on, mates, let's heave and haul_
> _We've got a new `BasePredictor` for us all_
> _It uses the backbone `embedding` for each task_
> _And makes our models smarter and more fast_

### Walkthrough
*  Add `embedding` argument to `postprocess` method of `BasePredictor` and its subclasses to use the output of the backbone network for different tasks ([link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-9fa2bf7f1273283680da3daeb8bae8e20cf8d477f838575796d2bf3b415bc534L188-R188), [link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-663b5aaaf7a04b1a7fe40326af039d2baf50df166e9738332eacbae9271ec55fL40-R40), [link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-9dded00f5e51a32b00c6266aa445990bae22fe6ea0bbb9d2c8898f4fcdb8cb23L23-R25), [link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-383afc92a2a66556c136cd357f28db596eed25b3472abe26b7e778208a365571L31-R31), [link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-a6dc03b08ac83fa9cfd1e48cbc1f4744b72e4fff1fe523df8936c0a144b50498L28-R28))
*  Modify `stream_inference` method of `BasePredictor` to unpack the output of `inference` method into `preds` and `embedding`, and pass `embedding` to `postprocess` method ([link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-9fa2bf7f1273283680da3daeb8bae8e20cf8d477f838575796d2bf3b415bc534L263-R267))
*  Add `embedding` argument to `__init__` method of `Results` class and assign it to `self.embedding` attribute to store the output of the backbone network for each task ([link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-24af401e53ffc5ce49197395e1e3d7111344ad31b08cbb2bee9377614da98955L92-R92), [link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-24af401e53ffc5ce49197395e1e3d7111344ad31b08cbb2bee9377614da98955R105))
*  Modify `postprocess` method of each subclass of `BasePredictor` to pass `embedding` argument to `Results` constructor ([link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-663b5aaaf7a04b1a7fe40326af039d2baf50df166e9738332eacbae9271ec55fL49-R49), [link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-9dded00f5e51a32b00c6266aa445990bae22fe6ea0bbb9d2c8898f4fcdb8cb23L40-R41), [link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-383afc92a2a66556c136cd357f28db596eed25b3472abe26b7e778208a365571L52-R52), [link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-a6dc03b08ac83fa9cfd1e48cbc1f4744b72e4fff1fe523df8936c0a144b50498L54-R54))
*  Import `torch.nn.functional` module in `ultralytics/nn/tasks.py` to use `adaptive_avg_pool2d` and `normalize` functions for processing the embedding output of the backbone network ([link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-e2650ac65b6525f6d9c0594a226066d396afbcba637c054d29abe04e62f410c4R9))
*  Add `_average_and_normalize` method to `BaseModel` class in `ultralytics/nn/tasks.py` to compute the average and normalize the 20x20 layers in the input tensor ([link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-e2650ac65b6525f6d9c0594a226066d396afbcba637c054d29abe04e62f410c4L60-R86))
*  Modify `_predict_once` method of `BaseModel` class in `ultralytics/nn/tasks.py` to initialize and assign `img_embedding` vector when the model iteration reaches a head class, and return both `x` and `img_embedding` as the output of the prediction ([link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-e2650ac65b6525f6d9c0594a226066d396afbcba637c054d29abe04e62f410c4L74-R109), [link](https://github.com/ultralytics/ultralytics/pull/6320/files?diff=unified&w=0#diff-e2650ac65b6525f6d9c0594a226066d396afbcba637c054d29abe04e62f410c4L83-R119))

### User Story
When sorting through photos I would like to quickly identify same or similar images using the image feature vector. I would like to do this on a single model pass when employing Detect, Segment, Class, Pose, etc.  

Otherwise, I would have to Run the selected Task (Detect, etc) and then Run the same image through the model without the Head ie) `emb_model.model.model = emb_model.model.model[:-1]`.  

This results in duplicate model runs for the same image, increasing processing time and overhead.  Also, requiring two models on the GPU; Normal Full Model and Embedding Model, using up valuable resources.

## Proposed Solution
Using the provided YOLOv8 models and functions, capture and save the backbone vector as the feature embedding for the entire image, pass the vector to the post-processing function and return the embedding in the Results class along with the predictions, probs, etc for display to the user.  Furthermore, allowing user to iterate over the results in a list of vectors; same as probs, boxes, etc.

### Sample Result Output

With the 'embeddding' item in the Results output.  When provided a batch of images, `embedding` returns a list of list of embeddings for each image.

```bash
[ultralytics.engine.results.Results object with attributes:
 
 boxes: ultralytics.engine.results.Boxes object
 embedding: [-0.007253803312778473, 0.03278650343418121, -0.008058142848312855, ...truncated... ]
 keypoints: None
 masks: None
 names: {0: 'person', 1: 'bicycle', 2: 'car', 3: 'motorcycle', 4: ..truncated..}
 orig_img: array([[[119, 146, 172],
         [121, 148, 174],
         [122, 152, 177],
         ...truncated....
```

## Related work in PRs, Issues, Branches

- PR #3094 extends the YOLOv8 classes to visualize the feature space of a given set of images, but does not extend the Results class to pass vector to user on model results
- Issue #1724 Requests a method to extract features
- Issue #3957 Requests feature extraction as well.

I propose this PR in order to extend the YOLOv8 code base to all users to extract the image features from the backbone, while also addressing the above requested issues and branches.

## Limitations
This method will only provide the embedding of the full input image.  It will NOT provide embeddings of crops, segments, etc.  For embeddings of detection crops, it will require the user to do image transformations and send crops through the model again to retrieve the crop embeddings.

The latent space of the full image embedding cannot be reduced or subset to fit each crop.

## Testing
Included is `embed_test.ipynb` for running the embeddings on each Task using the standard asset photos for a single image or multiple images as inputs, verifying output is aligned with YOLOv8 construct.

No notable speed change in detection or model processing time.

**License**: I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced YOLO models to output full image embeddings for advanced analysis.

### 📊 Key Changes
- New file `embed_test.ipynb` added demonstrating how to modify YOLOv8 to obtain image embeddings.
- Updated the model's prediction functionality to include image embeddings.
- Code adjustments in several files to handle embedding extraction across different model types.

### 🎯 Purpose & Impact
- **Purpose**: To enable users to get embeddings of the entire image from YOLO models, which are useful for feature comparison and advanced machine learning tasks.
- **Impact**: Broadens the applicability of YOLO models by integrating image embeddings, offering deeper insights and enhanced functionality for developers and researchers. 🧠🔍

### CODE REVIEW

**Comment by davidamacey** [comment]
I have read the CLA Document and I sign the CLA

**Comment by davidamacey** [comment]
I will follow-up with more testing to ensure the PR passes the unit tests and additional model support.

**Comment by glenn-jocher** [comment]
@davidamacey thank you for your contribution and for signing the CLA. 

It's great to hear that you're planning to conduct more testing, as thorough testing is crucial for ensuring the stability and reliability of new features. Once you have completed your tests and are confident in the PR's performance and compatibility, please ensure that the changes do not adversely affect any existing functionalities and that they align with the overall design principles of YOLOv8.

We highly value community feedback and contributions that enhance the capabilities of YOLO models, and your effort to address the feature extraction requests is appreciated. Before finalizing your PR, consider reviewing the documentation at https://docs.ultralytics.com to guarantee consistency and clarity in how the new feature should be used by the end-users.

Your work helps make YOLOv8 an even more versatile tool for the community. We're looking forward to reviewing your complete testing results and further discussions on this feature enhancement. Thank you for your dedication to improving the YOLOv8 project.

**Comment by davidamacey** [comment]
@glenn-jocher I have been reviewing the errors in the pytest.  They all seemed to be related to the cli and training tests.  

After some research of the code base and function outputs, it appears that the `forward` function calls the `_predict_once` function in 'ultralytics/nn/tasks.py', resulting in the forward pass outputs.  These outputs provide the specific results parameters, it is not always a single set of values.  It appears they are lists or tuples based on the model selected, ie SAM is a tuple of tensors to be post-processed.

The `forward` function is called for train and inference tasks, therefore my method of providing the 'preds' and 'embeddings' is not consistent with all models and all tasks.  It works for the basic YOLOv8 models when only using the Inference task.  Otherwise, the train forward pass outputs the preds and embedding, which conflicts with the necessary outputs for those tasks only requiring the preds.

I have done my best to understand the code base, but I could be missing something simple to allow me to pass the embeddings from the backbone to the Results Class. It is possible @AyushExel knows a better method with the related work on the 'embeddings' branch.

Any assistance or guidance would be greatly appreciated, as I believe feature embedding output would be useful.  Thank you!

**Comment by glenn-jocher** [comment]
@davidamacey, it's great to see your dedication to the enhancement of YOLOv8 and your proactive approach towards resolving the test errors.

You've clearly done a deep dive into the codebase and its functionalities. The challenge you’re facing with the `forward` function's variability in output structures across different tasks and models is a common issue when extending the capabilities of a complex system like YOLOv8.

The different output structures are indeed task-specific and are designed to support a wide range of tasks with varying output requirements. It’s also important to ensure backward compatibility, where newer changes do not break existing functionalities that users rely on.

For integrating the embeddings consistently across different models and tasks, you may need to abstract the embeddings in such a way that they can be optionally included in the outputs without disrupting the existing structure. This could potentially be implemented with a flag or setting at prediction time that determines whether embeddings should be calculated and included in the results.

Additionally, exploring the work being done on relevant branches, such as the 'embeddings' branch you have mentioned, is a good idea. Collaborating with other contributors who have worked on similar features can provide insights and potential strategies for integration that are consistent with the codebase's architecture.

If AyushExel or others have already explored this direction, reaching out within the existing issues or pull request discussions for advice could facilitate the process. Contributors with a history of working on similar areas might offer guidance or have suggestions that could resolve the compatibility conflicts with certain tasks.

Thank you once again for your effort in enriching the YOLOv8 project, and please keep us updated on your progress. We are here to support you in this endeavor, and together we can work towards a solution that fulfills the need for feature embeddings without compromising the framework's diverse functionalities.

**Review by ambitious-octopus — COMMENTED** [review]
Probably without a flag to pass during inference, we would have a significant overhead, but the idea is still interesting. What do you think @Laughing-q ? Is it worth spending some time on?

**Comment by Laughing-q** [comment]
@ambitious-octopus Sorry buddy I just saw this comment...I think we already this feature that returns embedding directly i.e. output the embedding from 22th layer:
```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")
embedding = model.predict(embed=[22])
```
@davidamacey Thanks for the PR! I think I'm going to close this one since we already have a similar feature. :)

**Comment by GabiShahmayster** [comment]
@Laughing-q 
Hi,
First of all, thank you for the wonderful work.
Could you please provide a simple example of obtaining the embedding for a number of detected objects, i.e the embedding per object, for downstream similarity search?
Thank you!
Gabi

**Comment by davidamacey** [comment]
@Laughing-q To get the embedding from each object would require an additional forward pass on each object crop.  Based on the current framework it is not possible to process the entire image with embedding, receive all bounding boxes, then provide bounding box crop embeddings.

Sending the resulting batches of crops through the forward pass gets complicated since not all crops are the same size.  Padding images adds blank buffer edges which could distort the comparison to the same object when its a different size in a different image.  YOLO has padding and image buffers, but careful preprocessing is required to get true image crop comparisons.

**Comment by gigumay** [comment]
Hi @davidamacey, 
thanks for your work - just stumbled across it. You say "The latent space of the full image embedding cannot be reduced or subset to fit each crop." What do you base this on? Have you done any experiments? Just curious since I was thinking of trying something along those lines and if you have looked at this already it would save me some work.
